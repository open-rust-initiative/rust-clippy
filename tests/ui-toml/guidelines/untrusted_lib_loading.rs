#![allow(unused)]
#![warn(clippy::untrusted_lib_loading)]
#![feature(rustc_private)]

extern crate libloading;
use std::ffi::{c_char, c_int, c_void, CString};
use std::io::Read;

extern "C" {
    fn scanf(format: *const c_char, buf: *mut c_char) -> i32;
    fn dlopen(f: *const c_char, flag: c_int) -> *mut c_void;
}

fn untrusted_io() -> String {
    "bad".to_string()
}

fn load_dylib(_lib: &str) -> i32 {
    0
}

mod lib_loader {
    pub fn load(_lib: &str) -> i32 {
        0
    }
}

fn load_using_libloading_and_custom_io() {
    use libloading::Library;

    let name_a = untrusted_io();
    let name_b = &untrusted_io();

    unsafe {
        let _a = Library::new(name_a); // lint
        let _b = libloading::Library::new(name_b); // lint
    }
}

fn load_using_dlopen_and_custom_io() {
    let name_a = untrusted_io().as_ptr() as *const c_char;
    let name_b = untrusted_io();

    unsafe {
        let _a = dlopen(name_a, 1); // lint
        let _b = dlopen(name_b.as_ptr() as *const c_char, 1); // lint
    }
}

fn custom_loaders() {
    use lib_loader::load;

    let name_a = std::fs::read_to_string("foo").unwrap();

    let _a = load_dylib(&name_a); // lint
    let _b = load(&name_a); // lint
    let _c = lib_loader::load(&name_a); // lint
}

fn main() {}
