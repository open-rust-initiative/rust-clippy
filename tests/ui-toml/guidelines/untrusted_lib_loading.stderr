error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:35:18
   |
LL |         let _a = Library::new(name_a); // lint
   |                  ^^^^^^^^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:31:18
   |
LL |     let name_a = untrusted_io();
   |                  ^^^^^^^^^^^^^^
   = note: `-D clippy::untrusted-lib-loading` implied by `-D warnings`

error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:36:18
   |
LL |         let _b = libloading::Library::new(name_b); // lint
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:32:19
   |
LL |     let name_b = &untrusted_io();
   |                   ^^^^^^^^^^^^^^

error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:45:18
   |
LL |         let _a = dlopen(name_a, 1); // lint
   |                  ^^^^^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:41:18
   |
LL |     let name_a = untrusted_io().as_ptr() as *const c_char;
   |                  ^^^^^^^^^^^^^^

error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:46:18
   |
LL |         let _b = dlopen(name_b.as_ptr() as *const c_char, 1); // lint
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:42:18
   |
LL |     let name_b = untrusted_io();
   |                  ^^^^^^^^^^^^^^

error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:55:14
   |
LL |     let _a = load_dylib(&name_a); // lint
   |              ^^^^^^^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:53:18
   |
LL |     let name_a = std::fs::read_to_string("foo").unwrap();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:56:14
   |
LL |     let _b = load(&name_a); // lint
   |              ^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:53:18
   |
LL |     let name_a = std::fs::read_to_string("foo").unwrap();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: loading dynamic library from untrusted source
  --> $DIR/untrusted_lib_loading.rs:57:14
   |
LL |     let _c = lib_loader::load(&name_a); // lint
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: untrusted IO function called here
  --> $DIR/untrusted_lib_loading.rs:53:18
   |
LL |     let name_a = std::fs::read_to_string("foo").unwrap();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 7 previous errors

